---
sidebar_position: 25
---

# u53-58 新知和技術 part2


## Promise：
1. 字面上的意思就是向某個對象說出一個你一定會做的事情，而當你做完事情就會通知該對象，在電腦科學裡，就是在程式B定義程式A處理完成之後便會根據處理結果來通知程式B，讓它做後續的處理。

2. promise vs. callback 比較起來，promise帶來幾點好處：
  - 擁有和callback可以幫助程式模組來回報處理結果的能力。
  - promise 可以自由將非同步處理打造成巢狀結構或者將非同步處理擺放在同一層，而callback只能以巢狀結構來打造，這對於人類而言，是相當易讀易開發-人類擅於處理從上至下排列的資訊，不擅於從外至內的巢狀資訊結構
  - callback 本身是賦予其他程式模組來執行後續結果如何處理，而promise本身是建立一個worker thread來去監聽其他程式模組是否完成執行，當完成時就會回復該worker thread並讓main thread來執行後續結果如何處理，前者對於是處理結果的函式-callback來說，callback本身執行次數會由其他程式模組來決定，後者則是完全是由main thread來決定次數，若把處理結果轉換成跟金流有關的處理，這問題會變得相當嚴重，比如付款處理會由其他程式模組重複付款。

3. 建立該機制必須要先建立Promise實體本身、判定事件的基準、前兩者的合併、對應結果本身的事件綁定：
4. 建立Promise 實體本身，其中executor是 function(resolve, reject)，而resolve和reject本身實現通知handle結果的函式，而function本身就是定義Promise所要答應要做的事情或者handle，過程中會添加resolve或者reject，當在裡面呼叫resolve時，就表示handle是成功的，並且會觸發對應的事件處理，而當在裡面呼叫reject時，就表示handle是失敗，同樣地會呼叫另一個事件處理

```
let promise = new Promise(executor)
// executor =>  function(resolve, reject)
```
另外當promise被定義的時候，就會等同執行promise內部的executor，另外可以像是goto的標籤那樣來分類哪些後續處理是屬於哪個promise-後續處理的程式碼會依賴於最近的promise本身或者類似標籤，這些標籤本身不會執行promise內容，只是純粹分類，比如：原本有兩段負責處理後續事件處理的程式碼，其中事件處理1是基於promise1來運作，事件處理2是基於promise2來運作。
```
// 後續事件處理1
// 後續事件處理2
```

為了進一步分類事件處理是屬於哪些promise1，這樣的效果會是後續事件處理1會依賴著promise1，而後續事件處理2會依賴著promise2

```
promise1
// 後續事件處理1

promise2
// 後續事件處理2
.
.
.
```

5. 判定事件的基準：Promise具有三種狀態來幫助系統判別每個Promise目前處於什麼狀態，並且把狀態當做事件以及進一步做事件處理，三種狀態分別為pending、fulfilled、rejected，在開始執行Promise時的狀態會是等待狀態(pending)，該狀態不會是fulfilled與rejected，等到Promise執行到resolve時，狀態就便會從 pending 轉換至 fulfilled，若是執行到reject時，就會從 pending 轉換至 rejected，後續事件判定就會根據這狀態來判別。

6. 而在Promise中的resolve和reject除了會把狀態轉換以外，還會將Promise的實際處理結果進行更改，比如說resolve(something)，就會把原本實際處理結果為undefined更換為something，而reject(error)則是把實際處理結果本實際處理結果為undefined更換為代表錯誤的error object。

```
let promise = new Promise(function(resolve, reject) {

  resolve(something)
  reject(error)
})
```

所以實際上來說當在Promise裡執行到resolve(something)，便會將狀態和實際處理結果分別轉換為fulfilled和something，而執行到reject(error)，便會將狀態和實際處理結果分別轉換rejected和error object，另外resolve和reject並不會執行return或者中斷後續執行。
![](https://res.cloudinary.com/dqfxgtyoi/image/upload/v1637749757/blog/promise/promiseTransfer_o3lnvo.png)


7. Promise本身有then、catch、finally這三個方法來定義promise被完成後的後續處理，這三個本身皆為不含有resolve和reject參數的promise，且這些promise將前一個promise的結果和狀態納入至自己的參數來做自己的promise處理：
```
// version 1：.then promise會接收promise1回傳的狀態和結果值來執行
promise1
  .then(...)

// 同等於
promise1.then(...)

// version 2 ：.catch promise會接收promise1回傳的狀態和結果值來執行
promise1
  .catch(...)

// 同等於
promise1.catch(...)

// version 3：.finally promise會接收promise1回傳的狀態和結果值來執行
promise1
  .finally(...)

// 同等於
promise1.finally(...)

```


8. .then會是定義fulfilled或者rejected所要有的事件處理，主要格式為如下，其中onFulfilled和onRejected為接收一個參數的函式，onFulfilled是負責接收前一個為fulfilled狀態的promise，並從中取出結果值當fullfillment參數執行onFulfilled函式本身，而onRejected是負責接收前一個rejected狀態的promsie，並從中取結果值當rejection參數執行onRejected函式本身，另外onRejected是可選擇性不添加。
```
.then(onFulfilled, onRejected)

// inside onFulfilled function
function onFulfilled(fullfillment) {

}

// inside onFulfilled function
function onRejected(rejection) {

}
```

而then最後會根據是否拋出錯誤來回傳xx狀態的promise給下一個.then/catch/finally，若目前then包含的程式碼未拋出錯誤，就回傳fulfilled狀態的promise，結果值會是依據return來決定，若拋出錯誤的話，就回傳rejected狀態的promise：
```
// version1 ：夾帶結果值為10的promise
.then(() => {
  return 10 
})
// version2 ：系統會在沒添加{}的function內，自動添加return，所以會是夾帶結果值為10的promise
.then(() => 10)

// version3 ：由於沒有return和添加{}，依照return結果來看會因爲沒回傳值而結果是undefined，
// 所以會是夾帶結果值為undefined且狀態為fulfilled的promise
.then(() => {
  10
})
```


9. .catch則是單方面接收前一個為rejected狀態的promise，而它的參數只會有onRejected這只接收一個參數的函式，當前一個promise的狀態為rejected時，便會從中接收前一個promise的結果值。
```
Promise.catch(onRejected);
// inside onFulfilled function
function onRejected(rejection) {

}
```
若 onRejected 發出錯誤的話，則 catch() 回傳的 Promise 所擁有的狀態是rejected，其他情形則是fulfilled。
```
// catch內發生錯誤，其結果會是failure value那段，並且回傳rejected狀態的promise
promise1
  .catch(() => {
    throw new Error()
  })
  .then((result) => {
    console.log('success value ', result)
  }, (error) => {
    console.log('failure value ', error)
  })

// 其他情形(沒發出任何錯誤)，其結果會是success value那段，並且回傳fulfilled狀態的promise
promise1
  .catch(() => {
    return value
  })
  .then((result) => {
    console.log(result)
  })
```

10. .finally則是定義無論前一個promise的狀態為何都皆會執行的事件處理，而它的參數只會是一個函式物件-onFinally，
```
Promise.finally(onFinally);

// inside onFinally function

function onFinally() {

}
```
當onFinally出現錯誤時，會回傳rejected的promise，否則皆會直接傳遞上一個promise的狀態和結果值
```
// version 2 ：finally發出錯誤的情況下
const promise1 = new Promise((resolve, reject) => {

  console.log('start')
  resolve(10)   // 變動狀態和改變結果值
  console.log('end')
})

// 由於finally內部是發出錯誤，所以會直接回傳rejected狀態的promise和錯誤訊息
promise1
  .finally(() => {
    throw new Error()
  })
  .then((result) => {
    console.log('success value ', result)
  }, (error) => {
    console.log('failure value ', error)
  })


// version 2 ：finally正常執行的情況下

const promise1 = new Promise((resolve, reject) => {

  console.log('start')
  resolve(10)   // 變動狀態和改變結果值
  console.log('end')
})

// 由於finally內部是正常執行，所以會直接將promise1的狀態和結果傳遞finally之後的then
promise1
  .finally(() => {
    return value
  })
  // 其結果會是promise1的狀態和結果值-fulfilled 和 10
  .then((result) => {
    console.log('success value ', result)
  }, (error) => {
    console.log('failure value ', error)
  })
```



## promise.all
1. 將多個promise的物件統整成一個promise來看待，若全部的promise都回傳resolve，那麼promise.all就回傳resolve，而若其中一個promise回傳reject，那麼promise.all就回傳reject
2. 格式為如下，其中promise1至promiseN為統整成一個promise的一部分。
```
promise.all(promise1, promise2, promise3, .....)
```
3. 硬性規定：
  - promise1, promise2, promise3,... 要透過return來回報自己的promise狀態和結果值，才能正常讓promise.all運作，否則若有promise沒回報，很有可能因為該promise回傳的結果是pending的promise而使得promise.all無限等待，進而停止運作，發出錯誤訊息。比如如下：
```
  Promise.all(
    users.map((user, user_index) => {
      return UserModel.create({ ...user }).then(...)
  )
```
4. 應用場景為多個promise同時間要處理時，可以使用
5. 好處：
  - 比單純建立多個promise來執行更有維護性、易讀性
  - 確保每個promise都能夠執行完，才執行指定需要先直接這些promise的任務


## Promise.resolve
1. Promise 的靜態方法，會直接回傳夾帶著fulfilled狀態和結果值value的Promise
```
Promise.resolve(value);
```
2. 主要作用於本身不存在resolve機制的promise結構，比如then、catch、finally

3. Promise.resolve vs. resolve：前者是建立fulfilled狀態的Promise，後者是輔助resolve所在的Promise來轉換fulfilled狀態並夾帶特定值的Promise

## async/await
1. async 和 await 都皆為promise相關的語法糖，主要是進一步讓promise在執行上是由上往下的執行方式，而非由外往內的執行方式
2. async 是搭配function的關鍵字，允許開發者在其函式中將以promise為主的非同步處理轉換為同步處理，並且會自動回傳夾雜著xx狀態和結果值的promise：
  - 若該function不會拋出錯誤和例外的話，xx狀態會是fulfilled，結果值會是回傳值，否則為undefined
  - 若該function拋出錯誤的話，xx狀態會是rejected，結果值會是錯誤訊息
3. await 是於async函式負責將promise為主的非同步處理轉換為同步處理，具體作法為如下，其中expression是對應著promise，而rv則是該promise的結果值(resolved value)
```
[rv] = await expression;
```

4. 使用時要把握三個原則:
  - 要先確認有 Promise 物件實例，也就是已經定義好 resolve/reject，才能使用 async/await
  - 在流程中正確設定關鍵字：async function 包裝起來、運用非同步處理的地方加上 await 關鍵字
  - 注意 async/await 和 then 不可以混搭使用，容易造成開發上的混亂。

5. 錯誤處理：由於await使得promise為主的非同步處理轉換成同步處理，因此可以使用專門處理同步處理的錯誤處理關鍵字try..catch
```
try {
  await expression1 
  await expression2
  .
  .
} catch(error) {
  console.log(error)
}
```

## 除錯練習
1. 

```
db.once('open', () => {
    // 對每個使用者進行增加和他們所擁有的餐廳
    users.forEach((user, user_index)=>{
        UserModel.create({
            ...user
        }).then((user) => {
            // 遍歷餐廳並且向資料庫系統發出增加餐廳的請求
            // 然而在這邊只是發送，並未有等待完成或者非同步的事件處理的機制
            restaurants.forEach((restaurant, rest_index)=>{
                if (rest_index >= 3*user_index && rest_index < 3*(user_index+1)) {
                    restaurant.userId = user._id
                    RestaurantModel.create(restaurant)
                }
            })
        // 當目前then事情做完之後，就接著執行下一個then
        }).then(()=>{
          // 確定目前使用者是否為最後一個，若是的話就關閉程式，否則就繼續執行
          // 由於是在增加餐廳的請求處理過程來進行離開，所以很有可能會有部分資料無法增加或者
          // 因為增加過程就關閉而報錯。
            console.log('done')
            if(user_index >= users.length-1) {
                process.exit()
            }
        })
    })
})
```

2. 

```
// 資料庫連線設定
// 資料庫連線完成
db.once('open', () => {
    new Promise((resolve, _reject)=>{
        for (const [user_index, user] of users.entries()) {
            UserModel.create({
                ...user
            }).then((user)=>{
                console.log('user created')
                const userRestaurant = []
                restaurants.forEach((restaurant, rest_index)=>{
                    if (rest_index >= 3*user_index && rest_index < 3*(user_index+1)) {
                        restaurant.userId = user._id
                        userRestaurant.push(restaurant)
                    }
                })
                return RestaurantModel.create(userRestaurant)

            // 等待目前使用者所擁有的餐廳資料建立完，就進入下一個then
            }).then(()=>{
              // 確定目前使用者是否為最後一個，若是的話就關閉程式，否則就繼續執行
                console.log('restaurant created')
                if(user_index >= users.length-1) {
                  // Promise.resolve本身是建立額外的fulfilled Promise
                  // 並不是以目前所在的Promise來表達resolve，而該Promise是在Promise chain的源頭
                  // 本身只能靠resolve和reject來回傳xx狀態的promise，所以下面這句會直接當成一般
                  // 程式碼來執行，不以resolve來執行
                    return Promise.resolve()
                }
            })
        }
    }).then(()=>{
        //等待所有使用者的餐廳資料創建完成
        console.log("所有使用者與餐廳資料創建完成")
        process.exit()
    })
})
```